extension/sigv4authextension/extension_test.go

```diff
@@ -0,0 +1,276 @@
+		},
+		{
+			name: "missing_region",
+			cfg: &Config{
+				Service: 	"service",
+				RoleArn:     	"rolearn",
+				STSEndpoint: 	"regional",
+			},
+			shouldError:   true,
+			expectedError: errNoRegionProvided.Error(),
+		},
+		{
+			name: "missing_service",
+			cfg: &Config{
+				Region:     	"region",
+				RoleArn:     	"rolearn",
+				STSEndpoint: 	"regional",
+			},
+			shouldError:   true,
+			expectedError: errNoServiceProvided.Error(),
+		},
+		{
+			name: "invalid_endpoint",
+			cfg: &Config{
+				Region:     	"region",
+				Service: 	"service",
+				RoleArn:     	"rolearn",
+				STSEndpoint: 	"stsendpoint",
+			},
+			shouldError:   true,
+			expectedError: errInvalidSTSEndpoint.Error(),
+		},
+	}
+
+	for _, testcase := range tests {
+		t.Run(testcase.name, func(t *testing.T) {
+			sa, err := newSigv4Extension(testcase.cfg, "runtimeInfo", zap.NewNop())
+			if testcase.shouldError {
+				assert.NotNil(t, err)
+				assert.Contains(t, err.Error(), testcase.expectedError)
+				return
+			}
+			assert.NoError(t, err)
+			assert.Equal(t, testcase.cfg.Region, sa.cfg.Region)
+			assert.Equal(t, testcase.cfg.Service, sa.cfg.Service)
+			assert.Equal(t, testcase.cfg.RoleArn, sa.cfg.RoleArn)
+			assert.Equal(t, testcase.cfg.STSEndpoint, sa.cfg.STSEndpoint)
+		})
+	}
+}
+
+type ErrorRoundTripper struct{}
+
+func (ert *ErrorRoundTripper) RoundTrip(r *http.Request) (*http.Response, error) {
+	return nil, errors.New("error")
+}
+
+func TestRoundTrip(t *testing.T) {
+	defaultRoundTripper := (http.RoundTripper)(http.DefaultTransport.(*http.Transport).Clone())
+	errorRoundTripper := &ErrorRoundTripper{}
+
+	tests := []struct {
+		name        	string
+		rt          	http.RoundTripper
+		shouldError 	bool
+		cfg  		*Config
+	}{
+		{
+			"valid_round_tripper",
+			defaultRoundTripper,
+			false,
+			&Config{Region: "region", Service: "service"},
+		},
+		{
+			"error_round_tripper",
+			errorRoundTripper,
+			true,
+			&Config{Region: "region", Service: "service", RoleArn: "rolearn"},
+		},
+	}
+
+	runtimeInfo := "runtimeInfo"
+
+	for _, testcase := range tests {
+		t.Run(testcase.name, func(t *testing.T) {
+			server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				_, err := v4.GetSignedRequestSignature(r)
+				assert.NoError(t, err)
+				assert.Equal(t, runtimeInfo, r.Header.Get("User-Agent"))
+				w.WriteHeader(200)
+			}))
+			defer server.Close()
+			serverURL, _ := url.Parse(server.URL)
+
+			sa, _ := newSigv4Extension(testcase.cfg, runtimeInfo, zap.NewNop())
+			rt, err := sa.RoundTripper(testcase.rt)
+
+			assert.NoError(t, err)
+			req, err := http.NewRequest("POST", serverURL.String(), strings.NewReader(""))
+			assert.NoError(t, err)
+			res, err := rt.RoundTrip(req)
+			if testcase.shouldError {
+				assert.Nil(t, res)
+				assert.Error(t, err)
+				return
+			}
+			assert.NoError(t, err)
+			assert.Equal(t, res.StatusCode, 200)
+		})
+	}
+}
+
+func TestRoundTripper(t *testing.T) {
```

Comment:
Not too sure what this is testing, perhaps can make the name more descriptive

Reply:
Here I'm testing the `RoundTripper()` method and that it's returning the expected `SigningRoundTripper` struct; as for the name, I made test as 1:1 to functions/methods, so the names reflect the function/method they are testing. I can still change it if that's preferred instead

Reply:
Ah this may be related to my other point that it should be extension construction that fails when config is invalid, not `RoundTripper`. Then you wouldn't need to test `RoundTripper` as there isn't any behavior, just copying of fields. We want tests for logic that effects callers / users but not just implementation details like struct copying

Reply:
Would it be better to just remove this test then?